intern/openmesheffect/blender/intern/mfxCallbacks.cpp

```diff
@@ -82,42 +105,96 @@ OfxStatus before_mesh_get(OfxHost *host, OfxMeshHandle ofx_mesh) {
+  blender_point_count = blender_mesh->totvert;
+  blender_loop_count = 0;
   for (int i = 0 ; i < blender_mesh->totpoly ; ++i) {
     int after_last_loop = blender_mesh->mpoly[i].loopstart + blender_mesh->mpoly[i].totloop;
-    vertex_count = max(vertex_count, after_last_loop);
+    blender_loop_count = max(blender_loop_count, after_last_loop);
+  }
+  blender_poly_count = blender_mesh->totpoly;
+  blender_loose_edge_count = 0;
+  for (int i = 0 ; i < blender_mesh->totedge ; ++i) {
+    if (blender_mesh->medge[i].flag & ME_LOOSEEDGE)
+      ++blender_loose_edge_count;
   }
-  face_count = blender_mesh->totpoly;
 
-  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropPointCount, 0, point_count));
-  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropVertexCount, 0, vertex_count));
-  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropFaceCount, 0, face_count));
+  // figure out input geometry size on OFX side
+  ofx_point_count = blender_point_count;
+  ofx_vertex_count = blender_loop_count;
+  ofx_face_count = blender_poly_count;
+  ofx_no_loose_edge = (blender_loose_edge_count > 0) ? 0 : 1;
+  ofx_constant_face_count = (blender_loose_edge_count > 0 && blender_poly_count == 0) ? 2 : -1;
+
+  if (blender_loose_edge_count > 0) {
+    // turn blender loose edges into 2-vertex faces
+    ofx_vertex_count += 2*blender_loose_edge_count;
+    ofx_face_count += blender_loose_edge_count;
+    printf("Blender mesh has %d loose edges\n", blender_loose_edge_count);
+  }
+
+  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropPointCount, 0, ofx_point_count));
+  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropVertexCount, 0, ofx_vertex_count));
+  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropFaceCount, 0, ofx_face_count));
+  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropNoLooseEdge, 0, ofx_no_loose_edge));
+  MFX_CHECK(ps->propSetInt(&ofx_mesh->properties, kOfxMeshPropConstantFaceCount, 0, ofx_constant_face_count));
 
   // Define vertex colors attributes
   int vcolor_layers = CustomData_number_of_layers(&blender_mesh->ldata, CD_MLOOPCOL);
   char name[32];
   OfxPropertySetHandle vcolor_attrib;
   for (int k = 0; k < vcolor_layers; ++k) {
     sprintf(name, "color%d", k);
-    MFX_CHECK(mes->attributeDefine(ofx_mesh, kOfxMeshAttribVertex, name, 3, kOfxMeshAttribTypeUByte, &vcolor_attrib));
-    MFX_CHECK(ps->propSetInt(vcolor_attrib, kOfxMeshAttribPropIsOwner, 0, 0));
-    MLoopCol *vcolor_data = (MLoopCol*)CustomData_get(&blender_mesh->ldata, k, CD_MLOOPCOL);
-    MFX_CHECK(ps->propSetPointer(vcolor_attrib, kOfxMeshAttribPropData, 0, (void*)&vcolor_data[0].r));
-    MFX_CHECK(ps->propSetInt(vcolor_attrib, kOfxMeshAttribPropStride, 0, sizeof(MLoopCol)));
+
+    // Note: CustomData_get() is not the correct function to call here, since that returns
+    // individual values from an "active" layer of given type. We want CustomData_get_layer_n().
```

Comment:
I am not fluent with this API, what was the 'k' argument to `CustomData_get` for then?

Reply:
AFAICT, `CustomData_get(buffer, k, type)` returns pointer to `k`-th element of "active" layer of given type (ie. pointer for particular vertex, face, etc.), while `CustomData_get_layer_n(buffer, type, k)` returns pointer to `k`-th layer of given type (beginning of its array).

https://github.com/eliemichel/OpenMeshEffectForBlender/blob/f42b03dcf839429387392442a31b816f482c55ae/source/blender/blenkernel/BKE_customdata.h#L320-L342

Reply:
Oh right, I understand it the same way now that I read the docstring again indeed.